homeassistant/components/goodwe/sensor.py

```diff
@@ -190,3 +197,30 @@ def available(self) -> bool:
         return cast(GoodweSensorEntityDescription, self.entity_description).available(
             self
         )
+
+    async def async_reset(self, now):
+        """Reset the value back to 0 at midnight."""
+        self._previous_value = 0
+        self.coordinator.data[self._sensor.id_] = 0
```

Comment:
How is this intended to work?
The goodwe seems to return a `dict`, how does updating that dict help with resetting the inverter earlier?

Reply:
Because that dict is used as the state value in `def native_value(self):`, the `self.async_write_ha_state()` will ensure that state is written to the state machine (I think).
But I am still testing.

Reply:
I would suggest to do this midnight reset only for non battery inverters (which are unavailable at and now intentionally keep stale values). 
For hybrid inverter which never go to sleep this is probably unwanted.
i.e. do it only for inverter instances of 'DT' class

Reply:
... or maybe reset the value only if the entity is (really) unavailable ?

Reply:
> Because that dict is used as the state value in def native_value(self):, the self.async_write_ha_state() will ensure that state is written to the state machine (I think).

Yes, I can understand that's the intention, but what happens when the update coordinator is called again 10 seconds later?
I've marked this PR as draft until you have a solution which is tested and confirmed to work.

Reply:
The update coordinator will fail since the inverter is offline at night, and the previous value will be reported which is also set to 0 during reset, so the value will continue to be 0.

I now also included a check that the invertor is indeed offline when resetting (in case a house battery is connected and the inverter is reachable, the reset is not needed).

I have tested this code in my setup and it is confirmed to work.